<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.4"/>
<title>NVBIO: $title</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">NVBIO
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.4 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(11)"><span class="SelectionMark">&#160;</span>Groups</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(12)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="index.html">NVBIO</a></li>  </ul>
</div>
</div><!-- top -->
<div class="contents">
<div class="textblock"><p>nvBowtie is a GPU-accelerated re-engineering of Bowtie2, a very widely used short-read aligner. While being completely rewritten from scratch, nvBowtie reproduces many (though not all) of the features of Bowtie2.</p>
<h1><a class="anchor" id="nvBowtieArch"></a>
Architecture</h1>
<p>In order to take advantage of the massive parallelism available in modern processor architectures, nvBowtie re-implements the same underlying algorithms as Bowtie2 taking a fundamentally different approach. In fact, while Bowtie2 is essentially designed to operate on a single read at a time (possibly having multiple CPU threads working on a different read), carrying the entire alignment process in what is basically a very complex chain of nested function calls, nvBowtie works at all times with large batches of reads, and treats their alignment as a complex pipeline composed by many relatively simple but deeply parallel stages. In many of these stages parallelism is spread at a much finer granularity than at the reads-level, for example by processing many candidate hits from each read at the same time.</p>
<p>There are effectively a few different pipelines, one for each kind of alignment:</p>
<ul>
<li>best mapping of single-end reads</li>
<li>best mapping of paired-end reads</li>
<li>all mapping of single-end reads</li>
<li>all mapping of paired-end reads (not implemented yet)</li>
</ul>
<h2><a class="anchor" id="BestMapping-SingleEnd"></a>
best-mapping, single-end reads</h2>
<p>This pipeline can be described by the following pseudo-code:</p>
<div class="fragment"><div class="line">best-alignments = {}</div>
<div class="line"></div>
<div class="line">seed-mapping-queue[in] := <a class="code" href="namespacenvbio_1_1cuda.html#a5702be3dee6b758f4895f831e75a9c72">all</a> reads</div>
<div class="line"><span class="keywordflow">while</span> (seed-mapping-queue[in] not empty)</div>
<div class="line">{</div>
<div class="line">    (seed-mapping-queue[out], seed-deques) := <a class="code" href="group___mapping.html#ga1c1a5dc606aedfe249f958e15ac5791f">map</a>( seed-mapping-queue[in] );</div>
<div class="line"></div>
<div class="line">    score( seed-mapping-queue[in], seed-deques, best-alignments );          <span class="comment">// find the best alignments for each read</span></div>
<div class="line"></div>
<div class="line">    seed-mapping-queue[in] = seed-mapping-queue[out];</div>
<div class="line">}</div>
<div class="line"><a class="code" href="group___traceback_detail.html#ga2def93f93f54896e327da2c4adb35a72">traceback_best</a>( best-alignments );                                          <span class="comment">// traceback the best alignments</span></div>
</div><!-- fragment --><p>where score() is a sub-pipeline:</p>
<div class="fragment"><div class="line">score([in] input-reads, [in] seed-deques, [in/out] best-alignments)</div>
<div class="line">{</div>
<div class="line">   <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads := input-reads</div>
<div class="line">   <span class="keywordflow">while</span> (<a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads not empty)</div>
<div class="line">   {</div>
<div class="line">        (hits-queue, <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads) := <a class="code" href="group___basic_utils.html#gabc969419bf0a10c17e84ceedd624cd49">select</a>( <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads, seed-deques );  <span class="comment">// select the next round of hits to process</span></div>
<div class="line">        <a class="code" href="group___f_m_index.html#ga036b0d1d5e131f9ae61bf9fc37e4949f">locate</a>( hits-queue );                                               <span class="comment">// SA coordinates -&gt; linear genome coordinates conversion</span></div>
<div class="line">        <a class="code" href="group___scoring.html#ga56e68063eee00be8726c13bbb9d92556">score_best</a>( hits-queue );                                           <span class="comment">// assign a score to the selected hits using DP alignment</span></div>
<div class="line">        <a class="code" href="group___reduce.html#ga01103d0dd86260c207369ef1f0d5b067">score_reduce</a>( hits-queue, best-alignments );                        <span class="comment">// keep track of the best alignments found so far for each read</span></div>
<div class="line">   }</div>
<div class="line">}</div>
</div><!-- fragment --><p>The various functions involved correspond to different pipeline stages:</p>
<ul>
<li><a class="el" href="group___mapping.html">Mapping</a></li>
<li><a class="el" href="group___locate.html">Locate</a></li>
<li><a class="el" href="group___select.html">Select</a></li>
<li><a class="el" href="group___scoring.html">Scoring</a></li>
<li><a class="el" href="group___reduce.html">Reduce</a></li>
<li><a class="el" href="group___traceback.html">Traceback</a></li>
</ul>
<p>The data flowing through the pipeline goes mostly through a single-data structure of type ScoringQueues (<a class="el" href="group___scoring_queues_module.html">Scoring Queues</a>), which itself contains the active-reads (ScoringQueues::active_reads), the hits-queue (ScoringQueues::hits) and an index allowing to walk all the hits belonging to each read (ScoringQueues::hits_index).</p>
<h2><a class="anchor" id="BestMapping-PairedEnd"></a>
best-mapping, paired-end reads</h2>
<p>This pipeline can be described by the following pseudo-code:</p>
<div class="fragment"><div class="line">best-anchor-alignments   = {}</div>
<div class="line">best-opposite-alignments = {}</div>
<div class="line"></div>
<div class="line"><span class="keywordflow">for</span> (anchor in {mate1,mate2})</div>
<div class="line">{</div>
<div class="line">    seed-mapping-queue[in] := <a class="code" href="namespacenvbio_1_1cuda.html#a5702be3dee6b758f4895f831e75a9c72">all</a> reads</div>
<div class="line">    <span class="keywordflow">while</span> (seed-mapping-queue[in] not empty)</div>
<div class="line">    {</div>
<div class="line">        (seed-mapping-queue[out], seed-deques) := <a class="code" href="group___mapping.html#ga1c1a5dc606aedfe249f958e15ac5791f">map</a>( seed-mapping-queue[in] );</div>
<div class="line"></div>
<div class="line">        score( seed-mapping-queue[in], seed-deques, best-anchor-alignments, best-opposite-alignments );</div>
<div class="line"></div>
<div class="line">        seed-mapping-queue[in] = seed-mapping-queue[out];</div>
<div class="line">    }</div>
<div class="line">}</div>
<div class="line"><a class="code" href="group___traceback_detail.html#ga2def93f93f54896e327da2c4adb35a72">traceback_best</a>( best-anchor-alignments );</div>
<div class="line"><a class="code" href="group___traceback.html#ga38c7416da4253853eae783371d0b797e">opposite_traceback_best</a>( best-opposite-alignments );</div>
</div><!-- fragment --><p>where score() is a sub-pipeline:</p>
<div class="fragment"><div class="line">score([in] input-reads, [in] seed-deques, [in/out] best-anchor-alignments, [in/out] best-opposite-alignments)</div>
<div class="line">{</div>
<div class="line">   <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads := input-reads</div>
<div class="line">   <span class="keywordflow">while</span> (<a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads not empty)</div>
<div class="line">   {</div>
<div class="line">        (hits-queue, <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads) := <a class="code" href="group___basic_utils.html#gabc969419bf0a10c17e84ceedd624cd49">select</a>( <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads, seed-deques );  <span class="comment">// select the next round of hits to process</span></div>
<div class="line">        <a class="code" href="group___f_m_index.html#ga036b0d1d5e131f9ae61bf9fc37e4949f">locate</a>( hits-queue );                                               <span class="comment">// SA coordinates -&gt; linear genome coordinates conversion</span></div>
<div class="line">        <a class="code" href="group___scoring.html#ga25139a24703152f236d5b0d4dbaf4426">anchor_score_best</a>( hits-queue );                                    <span class="comment">// assign a score to the selected hits using DP alignment</span></div>
<div class="line">        <a class="code" href="group___scoring.html#ga30a0a305deae846f983459397f8617ed">opposite_score_best</a>( hits-queue );                                  <span class="comment">// assign a score to opposite mate of the selected hits using DP alignment</span></div>
<div class="line">        <a class="code" href="group___reduce.html#ga6b7b95482ecb303c4827ddb318515609">score_reduce_paired</a>( hits-queue, best-anchor-alignments, best-opposite-alignments ); <span class="comment">// keep track of the best alignments found so far for each read</span></div>
<div class="line">   }</div>
<div class="line">}</div>
</div><!-- fragment --><h1><a class="anchor" id="nvBowtieUsage"></a>
Usage</h1>
<p>At the moment, the command line options of nvBowtie differ from those of bowtie2. A comprehensive list can be obtained running: </p>
<pre class="fragment">*  ./nvBowtie --help
* 
*  nvBowtie [options] reference-genome read-file output
*   options:
*     General:
*       --max-reads        int [-1]      maximum number of reads to process
*       --device           int [0]       select the given cuda device
*       --file-ref                       load reference from file
*       --server-ref                     load reference from server
*       --phred33                        qualities are ASCII characters equal to Phred quality + 33
*       --phred64                        qualities are ASCII characters equal to Phred quality + 64
*       --solexa-quals                   qualities are in the Solexa format
*       --pe                             paired ends input
*       --ff                             paired mates are forward-forward
*       --fr                             paired mates are forward-reverse
*       --rf                             paired mates are reverse-forward
*       --rr                             paired mates are reverse-reverse
*       --verbosity                      verbosity level
*     Seeding:
*       --seed-len         int [22]      seed lengths
*       --seed-freq        int [15]      interval between seeds
*       --max-hits         int [100]     maximum amount of seed hits
*       --max-reseed       int [2]       number of reseeding rounds
*     Extension:
*       --rand                           randomized seed selection
*       --max-dist         int [15]      maximum edit distance
*       --max-effort-init  int [15]      initial maximum number of consecutive extension failures
*       --max-effort       int [15]      maximum number of consecutive extension failures
*       --min-ext          int [30]      minimum number of extensions per read
*       --max-ext          int [400]     maximum number of extensions per read
*       --minins           int [0]       minimum insert length
*       --minins           int [500]     maximum insert length
*       --overlap                        allow overlapping mates
*       --dovetail                       allow dovetailing mates
*       --no-mixed                       only report paired alignments
*     Reporting:
*       --mapQ-filter      int [0]       minimum mapQ threshold
* </pre><p>While most parameters should be easy to understand, a major difference is that nvBowtie relies on external programs to build the reference indices:</p>
<ul>
<li><em>nvBWT</em>: builds the BWT indices of the reference FASTA files</li>
<li><em>nvSSA</em>: builds the Sampled Suffix Arrays corresponding to a given BWT</li>
</ul>
<p>e.g. suppose you have the human genome in a single FASTA file, <em>hg19.fa</em>; the following commands will create all indices needed by nvBowtie:</p>
<pre class="fragment">*  ./nvBWT hg19.fa hg19-index
*  ./nvSSA hg19-index
* </pre><p>At this point, one can run nvBowtie:</p>
<pre class="fragment">*  ./nvBowtie --file-ref hg19 my_reads.fastq my_reads.bam
* </pre><p>Note the presence of the option <em>&ndash;file-ref</em>, specifying that the reference indices come from disk. Another noteworthy option is to let nvBowtie fetch them from a <em>shared memory</em> server which can be run in the background: the <em>nvFM-server</em>. It be launched with:</p>
<pre class="fragment">*  ./nvFM-server hg19-index hg19 &amp;
* </pre><p>At this point, nvBowtie can pickup the reference from the server:</p>
<pre class="fragment">*  ./nvBowtie hg19 my_reads.fastq my_reads.bam
* </pre> </div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Fri Jan 17 2014 10:59:41 for NVBIO by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.4
</small></address>
</body>
</html>
